src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -124,30 +135,43 @@ definition index :: "('i \<Rightarrow> 'q) \<Rightarrow> 'q \<Rightarrow> 'i lis
 definition first_intersection_point :: "('i \<Rightarrow> 'q) \<Rightarrow> 'q list \<Rightarrow> 'i list \<rightharpoonup> 'q" where
   [simp]: "first_intersection_point \<psi> qs C  = find (\<lambda>q. q \<in> \<psi> ` set C) qs"
 
+definition last_intersection_point :: "('i \<Rightarrow> 'q) \<Rightarrow> 'i list \<Rightarrow> 'i list \<Rightarrow> 'q" where
+  [simp]: "last_intersection_point \<psi> C C' = (ARG_MAX (\<lambda>q. index \<psi> q C) q. q \<in> \<psi> ` (set C \<inter> set C'))"
+
 corec server_program where
-  "server_program \<psi> C k b =
+  "server_program \<psi> C k b a =
     \<down> M. (partial_case M of
       Done \<Rightarrow>
         \<bottom> |
       Cont (FindIntersect qs) \<Rightarrow>
         (case first_intersection_point \<psi> qs C of
           None \<Rightarrow>
             \<up> Cont IntersectNotFound;
-            server_program \<psi> C k b |
+            server_program \<psi> C k b a |
           Some q \<Rightarrow>
             \<up> Cont (IntersectFound q);
-            server_program \<psi> C (index \<psi> q C) True
+            server_program \<psi> C (index \<psi> q C) True a
         ) |
       Cont RequestNext \<Rightarrow>
         if b then
           \<up> Cont (RollBackward (\<psi> (C ! k)));
-          server_program \<psi> C k False
+          server_program \<psi> C k False a
         else if Suc k < length C then
           \<up> Cont (RollForward (C ! Suc k));
-          server_program \<psi> C (Suc k) b
-        else \<comment> \<open>only for this initial implementation\<close>
+          server_program \<psi> C (Suc k) b a
+        else
           \<up> Cont AwaitReply;
-          server_program \<psi> C k b
+          a \<rightarrow> C'. (
+            \<comment> \<open>assumed that at least \<^term>\<open>C\<close>~and~\<^term>\<open>C'\<close> coincide in the genesis block\<close>
```

Comment:
Yes, there is no immediate way to formally specify these assumptions. That said, I think that we should document them next to the locale specification, since they are similar to the locale assumptions. I suggest to place the following block right after the locale specification:
```isabelle
text ‹
  We assume that the environment sends to \<^term>‹server_chain_updates› only non-empty lists whose
  first elements are equal to the first element of \<^term>‹initial_server_chain›.
›
```

Reply:
Fixed in 403ec57ef3b1e836c84ff5619a7f94bdf1a8c359, please check. Note that, since `server_chain_updates` was removed as part of 4b6488c47e1500f3f69dfab23af37680da18676a, the documentation of this assumption is placed accordingly.

Reply:
Hmm, `server_chain_updates` was not removed but rather `initial_server_chain`. Therefore, the documentation of said assumption should still be placed right after the locale specification. I’ll take care of this.

Reply:
Taken care of in a4270d9933cbee899d1417bae5bd5bf0715d1e12. :slightly_smiling_face: